[B+C] Add QuitReason enum to PlayerQuitEvent + PlayerKickEvent. Fixes BUKKIT-429

dualspiral opened the pull request:
Adds feature requested in BUKKIT-429. https://bukkit.atlassian.net/browse/BUKKIT-429
CraftBukkit PR: Bukkit/CraftBukkit#944

This change adds a QuitReason to the PlayerQuitEvent and PlayerKickEvent as an enum. When the event is called, CraftBukkit passes the reason for the disconnect to the events - and it'll fallback to QuitReason.UNKNOWN if the reason is not known.

Plugins can then access the reason the player quit through the event.getQuitReason() method. A fallback constructor remains in place, just in case another plugin is calling the event, this defaults to "QuitReason.UNKNOWN" or "QuitReason.KICKED" every time, dependant on the event called.

A small test plugin is available for this change, source code can be found at dualspiral/quittest@58af03d99ce28f75b83bb0bc8e5e690bd57c8e91

The motivation for this change arises from developing a plugin to customize quit messages dependent on the reason a player quit. Currently, the only method that I've seen that has done this is through scraping the console, and sometimes, it does not catch the disconnect reason. Embedding this in Bukkit and CraftBukkit will solve this issue, should it be allowed.

(This description has been altered (PlayerQuitEvent -> QuitEvent) to reflect the state of the PR)
### Status
- [ ] Waiting for author (@dualspiral)
- [ ] Waiting for PRH (@riking)
- [x] Waiting for core

mbax commented on src/main/java/org/bukkit/event/player/PlayerQuitEvent.java:
```diff
@@ -41,4 +59,43 @@ public HandlerList getHandlers() {
+    
+    public enum PlayerQuitReason {
+        UNKNOWN("disconnect.unknown"),
+        QUTTING("disconnect.quitting"),
```
QUTTING

mbax commented on src/main/java/org/bukkit/event/player/PlayerQuitEvent.java:
```diff
@@ -9,12 +9,21 @@
         this.quitMessage = quitMessage;
+        this.reason = PlayerQuitReason.UNKNOWN;
     }
-
```
Delete a line, then add it! What.

LB-- replied:
Probably change in whitespace on the line

jrtc27 replied:
Point is that means something's gone wrong...

dualspiral replied:
That's probably due to me squashing commits. There was a previous version of the constructor, which took the reason as a string rather than an enum. I then thought that I'd better not be lazy, and I took it out. I'll get that fixed up.

dualspiral commented:
The diffs don't show it very well, but my IDE was putting in whitespace on blank lines. I removed them (this was what was causing the remove/add blank line on line 17.)

This whitespace change makes it consistent with the rest of the file.

riking commented:
Fix the comment formatting, you need a space after the double slashes.
Line 14 should be:

```
// Retained...
```

not 

```
//Retained...
```

Also, while you're at it, double check to make sure that you're following all the style guidelines, and do a `git rebase -i 49690f9620` to `s`quash it all into 1 commit.

dualspiral commented:
Ok, I've addressed the comments, checked I was following the style guidelines (mostly re-aligning comments, putting the correct amount of whitespace, think I'm conforming now!) and squashed it all into one commit.

jb-aero commented:
Very nice, glad to see I inspired someone more motivated than myself :)

wasnertobias commented:
Can you please pull this commit? This is really necessary.

mbax commented:
Why not just call the enum QuitReason?

Also, the deprecated method should be deprecated not just commented.

jrtc27 commented on src/main/java/org/bukkit/event/player/PlayerQuitEvent.java:
```diff
@@ -41,4 +59,43 @@ public HandlerList getHandlers() {
+        QUITTING("disconnect.quitting"),
+        ENDOFSTREAM("disconnect.endOfStream"),
+        OVERFLOW("disconnect.overflow"),
+        GENERICREASON("disconnect.genericReason"),
```
Would `GENERIC_REASON` not be a better and clearer name for it?

mbax replied:
Yes it would. Same with END_OF_STREAM

jrtc27 commented on src/main/java/org/bukkit/event/player/PlayerQuitEvent.java:
```diff
@@ -41,4 +59,43 @@ public HandlerList getHandlers() {
+         */
+        public static PlayerQuitReason getByString(final String reason) {
+            for (PlayerQuitReason r : PlayerQuitReason.values()) {
+              if (r.getReasonString().equals(reason)) {
```
4 spaces per indent

jrtc27 commented on src/main/java/org/bukkit/event/player/PlayerQuitEvent.java:
```diff
@@ -41,4 +59,43 @@ public HandlerList getHandlers() {
+         * @param reason The disconnect reason as given by CraftBukkit
+         * @return The disconnect reason.
+         */
+        public static PlayerQuitReason getByString(final String reason) {
```
Use a static Map for this so you can just do `reasons.get(reason);`. Look at the other enums for how to do this (use `static { ... }`).

jrtc27 commented on src/main/java/org/bukkit/event/player/PlayerQuitEvent.java:
```diff
@@ -41,4 +59,43 @@ public HandlerList getHandlers() {
+    public enum PlayerQuitReason {
+        UNKNOWN("disconnect.unknown"),
+        QUITTING("disconnect.quitting"),
+        ENDOFSTREAM("disconnect.endOfStream"),
```
As below with `GENERICREASON`, would `END_OF_STREAM` not be better?

dualspiral commented:
Addressed all comments above. Will push required CraftBukkit changes to other PR in a moment.

dualspiral commented:
RE last commit, just removing an extra comma in a comment.

EdGruberman commented:
I would suggest considering splitting QuitReason out and placing it in org.bukkit directly.  Then, instead of having to ensure the quit strings match up with nms code over time, you could replace occurrences of the quit string in usage with the enum value itself and pass the enum reference through the methods.

Wolvereness commented:
The various quit reason things and getQuitReason() do not belong in Bukkit. Having an enum is one thing, but the strings should map to reasons in CraftBukkit.

EdGruberman commented:
@Wolvereness Keeping the strings in CraftBukkit makes sense, but why not getQuitReason in the PlayerQuitEvent in Bukkit? How else can plugins react to the various reasons?

Wolvereness commented:
@EdGruberman https://github.com/Bukkit/Bukkit/pull/726#issuecomment-11314486

> @Wolvereness Keeping the strings in CraftBukkit makes sense, but why not getQuitReason in the PlayerQuitEvent in Bukkit? How else can plugins react to the various reasons?

Sorry, I was specifically referring to line https://github.com/dualspiral/Bukkit/commit/dbf6550707945403bc1cedb6feb765f691307b40#L0R39 

``` JAVA
public static QuitReason getByString(final String reason) {
```

jrtc27 commented:
So make that private/protected?

EdGruberman commented:
@jrtc27 No, Wolvereness is indicating all the string association should be kept and done in CraftBukkit.  In fact, you could drop the strings altogether (keep passing them to the logging function) but only pass the enum value itself through the API.  The strings themselves are irrelevant to the API since an enum is being used.

jrtc27 commented:
Oh right.

dualspiral commented:
Bukkit now only contains the enum and the getQuitReason() event in PlayerQuitEvent (which returns a QuitReason enum value)

I've made an attempt at implementing @EdGruberman's suggestion, passing the enum through the API. I'm not completely sure if it's OK, but the CB PR has the changes.

dualspiral commented:
It makes perfect sense once someone points it out! I generally do that in my own codes too, so I only have to change one method instead of two - so I have no idea why I didn't here. Might as well do it!

dualspiral commented:
Implemented substitutions as @EdGruberman suggested:

(Message below from previous commit to show substitutions)

> ILLEGAL is an excellent prefix to leverage grouping similar things in the enum. Consider the following substitutions:
> 
> SPAM = ILLEGAL_CHAT_RATE
> FLYING_WHEN_NOT_PERMITTED = ILLEGAL_FLIGHT
> DROPPED_ITEMS_TOO_QUICKLY = ILLEGAL_DROP_RATE
> PROTOCOL_ERROR = ILLEGAL_PROTOCOL
> INVALID_CARRIED_ITEM = ILLEGAL_ITEM
> ILLEGAL_CHAR_IN_CHAT = ILLEGAL_CHAT_CHARACTER
> CHAT_TOO_LONG = ILLEGAL_CHAT_LENGTH
> MODDED_CLIENT = ILLEGAL_CLIENT
> ATTACK_NOT_PERMITTED = ILLEGAL_ATTACK
> 
> ANOTHER_LOCATION might be more clear as LOGIN_LOCATION
> SERVER_CLOSED might be more clear as SERVER_SHUTDOWN (I think of closed as too similar to locked in a sense.)
> 
> Just suggestions based on my personal preference. I really like like where you are going with all this!

ST-DDT commented:
awesome, this would help me a lot.

spaceemotion commented:
:+1:

pschichtel commented:
:+1:

mbax commented:
This is not a vote site. Please do not +1 or thumbsup PRs. That will not affect their state.

pschichtel commented:
other open source projects do this

jackklink commented:
Currently my plugin is able to get this information in a roundabout way. I monitor the console log for the disconect.X

http://dev.bukkit.org/server-mods/crashnotifier/

Would love to simplify my methods if you could implement this.

riking commented on src/main/java/org/bukkit/event/player/PlayerQuitEvent.java:
```diff
@@ -9,10 +10,18 @@
     private String quitMessage;
+    private QuitReason reason;
 
+    // Deprecated - people using this will obtain a quit reason of PlayerQuitReason.UNKNOWN
```
Javadoc actually has a @Deprecated tag.

``` java
/**
 * @deprecated new parameter included, QuitReason
 */
```

Or something like that.

dualspiral committed:
Allow plugins to obtain the reason a player disconnected. Adds BUKKIT-429
Adds a QuitReason enum, and adds getQuitReason() to the PlayerQuitEvent and PlayerKickEvent.
If a disconnect reason appears that does not feature in the enum, it falls back to "QuitReason.UNKNOWN" or "QuitReason.KICKED". A fallback constructor (marked as Deprecated) remains in place, just in case someone is calling the event from their plugins, this defaults to "QuitReason.UNKNOWN" or "QuitReason.KICKED".

dualspiral commented:
I've added this to PlayerKickReason too, but if you would rather have that in a separate PR, let me know. I figured it made sense to just do it here!

riking commented on src/main/java/org/bukkit/QuitReason.java:
~~You need a newline at the end of the file~~

turt2live commented:
@dualspiral If you are still interested in this please update it to the latest code base. This PR will also need to match the CONTRIBUTING.md guidelines and will need to be updated to resolve the concerns mentioned. 

Once the PR is updated the review process can continue.

dualspiral commented:
See last comment in Bukkit/CraftBukkit#944 for reasons for closing (essentially, I have no time to continue with this).
